lsat/interceptor.go

```diff
@@ -0,0 +1,292 @@
+package lsat
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"regexp"
+	"sync"
+	"time"
+
+	"github.com/lightninglabs/loop/lndclient"
+	"github.com/lightningnetwork/lnd/lnrpc/routerrpc"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/macaroons"
+	"github.com/lightningnetwork/lnd/zpay32"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	// GRPCErrCode is the error code we receive from a gRPC call if the
+	// server expects a payment.
+	GRPCErrCode = codes.Internal
+
+	// GRPCErrMessage is the error message we receive from a gRPC call in
+	// conjunction with the GRPCErrCode to signal the client that a payment
+	// is required to access the service.
+	GRPCErrMessage = "payment required"
+
+	// AuthHeader is is the HTTP response header that contains the payment
+	// challenge.
+	AuthHeader = "WWW-Authenticate"
+
+	// MaxRoutingFee is the maximum routing fee in satoshis that we are
+	// going to pay to acquire an LSAT token.
+	// TODO(guggero): make this configurable
+	MaxRoutingFeeSats = 10
+
+	// PaymentTimeout is the maximum time we allow a payment to take before
+	// we stop waiting for it.
+	PaymentTimeout = 60 * time.Second
+)
+
+var (
+	// authHeaderRegex is the regular expression the payment challenge must
+	// match for us to be able to parse the macaroon and invoice.
+	authHeaderRegex = regexp.MustCompile(
+		"LSAT macaroon='(.*?)' invoice='(.*?)'",
+	)
+)
+
+// Interceptor is a gRPC client interceptor that can handle LSAT authentication
+// challenges with embedded payment requests. It uses a connection to lnd to
+// automatically pay for an authentication token.
+type Interceptor struct {
```

Comment:
Good idea, this whitelist would also need to be enforced at the proxy though. As is, it's only aware of the service, and but not the individual methods of the service itself.

Reply:
One route to implement this would be a new caveat that signals calls which don't need tokens. Or we just extend the existing `freebie` system to examine the request and allow certain calls.

Reply:
I've added a PR to the proxy to allow (whitelist) certain calls.